src/java.base/share/classes/jdk/internal/foreign/SegmentBulkOperations.java

```diff
@@ -0,0 +1,271 @@
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package jdk.internal.foreign;
+
+import jdk.internal.misc.ScopedMemoryAccess;
+import jdk.internal.util.Architecture;
+import jdk.internal.vm.annotation.ForceInline;
+
+import java.lang.foreign.MemorySegment;
+
+/**
+ * This class contains optimized bulk operation methods that operate on one or several
+ * memory segments.
+ * <p>
+ * Generally, the methods attempt to work with as-large-as-possible units of memory at
+ * a time.
+ * <p>
+ * It should be noted that when invoking scoped memory access get/set operations, it
+ * is imperative from a performance perspective to convey the sharp types from the
+ * call site in order for the compiler to pick the correct Unsafe access variant.
+ */
+public final class SegmentBulkOperations {
+
+    private SegmentBulkOperations() {}
+
+    private static final ScopedMemoryAccess SCOPED_MEMORY_ACCESS = ScopedMemoryAccess.getScopedMemoryAccess();
+
+    // All the threshold values below MUST be a power of two and should preferably be
+    // greater or equal to 2^3.
+
+    // Update the FILL value for Aarch64 once 8338975 is fixed.
+    private static final long NATIVE_THRESHOLD_FILL = powerOfPropertyOr("fill", Architecture.isAARCH64() ? 10 : 5);
+    private static final long NATIVE_THRESHOLD_MISMATCH = powerOfPropertyOr("mismatch", 10);
+    private static final long NATIVE_THRESHOLD_COPY = powerOfPropertyOr("copy", 6);
+
+    @ForceInline
+    public static MemorySegment fill(AbstractMemorySegmentImpl dst, byte value) {
+        dst.checkReadOnly(false);
+        if (dst.length == 0) {
+            // Implicit state check
+            dst.checkValidState();
+        } else if (dst.length < NATIVE_THRESHOLD_FILL) {
+            // 0 <= length < FILL_NATIVE_LIMIT : 0...0X...XXXX
+
+            // Handle smaller segments directly without transitioning to native code
+            final long u = Byte.toUnsignedLong(value);
+            final long longValue = u << 56 | u << 48 | u << 40 | u << 32 | u << 24 | u << 16 | u << 8 | u;
+
+            int offset = 0;
+            // 0...0X...X000
+            final int limit = (int) (dst.length & (NATIVE_THRESHOLD_FILL - 8));
+            for (; offset < limit; offset += 8) {
+                SCOPED_MEMORY_ACCESS.putLong(dst.sessionImpl(), dst.unsafeGetBase(), dst.unsafeGetOffset() + offset, longValue);
```

Comment:
Now that I look again at this - I think all these calls should use `putXYZUnaligned`. Otherwise alignment can introduce issues on some platforms.

Reply:
This tactic should be reflected in a top-level comment in this class.